test/e2e/persistent_volumes.go

```diff
@@ -47,16 +47,158 @@ func nfsServerPodCleanup(c *client.Client, config VolumeTestConfig) {
 
 // Delete the PV. Fail test if delete fails.
 func deletePersistentVolume(c *client.Client, pv *api.PersistentVolume) {
-	// Delete the PersistentVolume
-	framework.Logf("Deleting PersistentVolume")
+
+	By("Deleting PersistentVolume")
+
+	framework.Logf("Deleting PersistentVolume %v", pv.Name)
 	err := c.PersistentVolumes().Delete(pv.Name)
 	if err != nil {
 		framework.Failf("Delete PersistentVolume %v failed: %v", pv.Name, err)
 	}
-	// Wait for PersistentVolume to Delete
+
+	// Wait for PersistentVolume to delete
 	framework.WaitForPersistentVolumeDeleted(c, pv.Name, 3*time.Second, 30*time.Second)
 }
 
+// Delete the PVC and wait for the PV to become Available again.
+// Validate that the PV has recycled (assumption here about reclaimPolicy).
+// Fail test if delete or wait fails.
+func deletePVCandValidatePV(c *client.Client, ns string, pvc *api.PersistentVolumeClaim, pv *api.PersistentVolume) {
+
+	By("Deleting PersistentVolumeClaim to trigger PV Recycling")
+
+	framework.Logf("Deleting PersistentVolumeClaim %v to trigger PV Recycling", pvc.Name)
+	err := c.PersistentVolumeClaims(ns).Delete(pvc.Name)
+	if err != nil {
+		framework.Failf("Delete PersistentVolumeClaim %v failed: %v", pvc.Name, err)
+	}
+
+	// Wait for the PV's phase to return to Available
+	framework.Logf("Waiting for recycling process to complete.")
+	err = framework.WaitForPersistentVolumePhase(api.VolumeAvailable, c, pv.Name, 3*time.Second, 300*time.Second)
+	if err != nil {
+		framework.Failf("Recycling failed: %v", err)
+	}
+
+	// Examine the pv.ClaimRef and UID. Expect nil values.
+	pv, err = c.PersistentVolumes().Get(pv.Name)
+	if err != nil {
+		framework.Failf("Cannot re-get PersistentVolume %v:", pv.Name, err)
+	}
+	if pv.Spec.ClaimRef != nil && len(pv.Spec.ClaimRef.UID) > 0 {
+		crJSON, _ := json.Marshal(pv.Spec.ClaimRef)
+		framework.Failf("Expected PV %v's ClaimRef to be nil, or the claimRef's UID to be blank. Instead claimRef is: %v", pv.Name, string(crJSON))
+	}
+}
+
+// create the PV resource. Fails test on error.
+func createPV(c *client.Client, pv *api.PersistentVolume) (*api.PersistentVolume) {
+
+	pv, err := c.PersistentVolumes().Create(pv)
+	if err != nil {
+		framework.Failf("Create PersistentVolume %v failed: %v", pv.Name, err)
+		return nil
```

Comment:
`return nil` suggests, that the execution after `Failf` continues, which is IMO not the case. Please remove it (and also few lines below in `createPVC`.